rz-svd/rz-svd.c

```diff
@@ -0,0 +1,309 @@
+#include <rz_analysis.h>
+#include <rz_cmd.h>
+#include <rz_cons.h>
+#include <rz_core.h>
+#include <rz_lib.h>
+#include <rz_types.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <yxml.h>
+
+#define XMLBUFSIZE 4096
+#define VALUESIZE  2048
+
+typedef struct ta_iter {
+	uint8_t yxml_buf[XMLBUFSIZE];
+	yxml_t x;
+
+	const char *start;
+	const char *ptr;
+	const char *end;
+
+	char regname[50];
+	char baseaddress[50];
+	char bitoffset[10];
+	char bitwidth[20];
+	char description[50];
+} ta_iter;
+
+enum { REGNAME,
+	BITWIDTH,
+	BITOFFSET,
+	DESCRIPTION } elem_type;
+
+static void strcpytrunc(char *dst, const char *src, size_t dstsize) {
+	size_t to_copy = strlen(src);
+	if (to_copy >= dstsize)
+		to_copy = dstsize - 1;
+	memcpy(dst, src, to_copy);
+	dst[to_copy] = '\0';
+}
+
+static const RzCmdDescArg cmd_svd_args[] = {
+	{
+		.name = "Path to the SVD file",
+		.type = RZ_CMD_ARG_TYPE_FILE,
+
+	},
+	{ 0 },
+};
+
+static const RzCmdDescHelp svd_usage = {
+	.summary = "SVD Plugin for Rizin",
+	.args = cmd_svd_args,
+};
+
+static inline int ta_iter_done(ta_iter *ta) {
+	return *ta->ptr == 0 || ta->ptr >= ta->end;
+}
+
+static ta_iter *ta_iter_next(ta_iter *ta) {
+	yxml_ret_t r = YXML_OK;
+	yxml_t ta_x;
+
+	int level;
+	char *cur, *tmp;
+	char value[VALUESIZE];
+	const char *ta_start;
+
+	cur = value;
+	value[0] = 0;
+
+	if (!ta->baseaddress[0]) {
+		while (!ta_iter_done(ta) &&
+			(yxml_parse(&ta->x, *ta->ptr) != YXML_ELEMSTART ||
+				strcasecmp(ta->x.elem, "peripherals"))) {
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+		ta_start = ta->ptr;
+		ta_x = ta->x;
+
+		level = 0;
+		while (!ta_iter_done(ta) && !ta->baseaddress[0]) {
+			switch ((r = yxml_parse(&ta->x, *ta->ptr))) {
+			case YXML_ELEMSTART:
+				level += 1;
+				if (level == 2 && strcasecmp(ta->x.elem, "baseAddress") == 0) {
+					cur = value;
+					*cur = 0;
+				}
+				break;
+
+			case YXML_ELEMEND:
+				level -= 1;
+				if (level < 0) {
+					return ta_iter_next(ta);
+				} else if (level == 2 && cur) {
+					strcpytrunc(ta->baseaddress, value, sizeof(ta->baseaddress));
+					cur = NULL;
+
+					// go back to the beginning of peripherals?
+					ta->ptr = ta_start;
+					ta->x = ta_x;
+				}
+				break;
+
+			case YXML_CONTENT:
+				if (!cur || level != 2) {
+					break;
+				}
+				tmp = ta->x.data;
+				while (*tmp && cur < value + sizeof(value)) {
+					*cur++ = *tmp++;
+				}
+				if (cur >= value + sizeof(value)) {
+					cur = NULL;
+				} else {
+					*cur = 0;
+				}
+				break;
+			default:
+				break;
+			}
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+	}
+	assert(ta->baseaddress[0]);
+
+	// find <register>
```

Comment:
Probably better to split this part into a separate function

Reply:
Done.